implementations/rust/ockam/ockam/src/entity/profile.rs

```diff
@@ -31,40 +69,258 @@ impl Profile {
         }
     }
 }
 
 impl Profile {
-    pub fn apply(&mut self, change_event: ProfileChangeEvent) {
-        if !self.verify(&change_event) {
-            return; // TODO: Throw error
+    pub fn create_with_key(
+        label: String,
+        key_type: ProfileKeyType,
+        key_purpose: ProfileKeyPurpose,
+        attributes: Option<ProfileEventAttributes>,
+        vault: Arc<Mutex<dyn ProfileVault>>,
+    ) -> Result<Self, ockam_core::Error> {
+        let attributes = attributes.unwrap_or(ProfileEventAttributes::new());
+
+        let mut v = vault.lock().unwrap();
+
+        // TODO: Should be customisable
+        let secret_attributes = SecretAttributes {
+            stype: SecretType::Curve25519,
+            persistence: SecretPersistence::Persistent,
+            length: 0,
+        };
+
+        let private_key = v.secret_generate(secret_attributes)?;
+        let public_key = v.secret_public_key_get(&private_key)?;
+
+        let change =
+            CreateKeyChange::new(label, key_type, key_purpose, public_key.as_ref().to_vec());
+        let prev_id = v.sha256(&[])?;
+        let prev_id = EventIdentifier::from_hash(prev_id);
+        let profile_change =
+            ProfileChange::new(1, prev_id, attributes, ProfileChangeType::CreateKey(change));
+        let changes = Changes::new(vec![profile_change]);
+        let changes_binary =
+            serde_bare::to_vec(&changes).map_err(|_| OckamError::BareError.into())?;
+
+        let event_id = v.sha256(&changes_binary)?;
+        let event_id = EventIdentifier::from_hash(event_id);
+
+        let self_signature = v.sign(&private_key, event_id.as_ref())?;
+        let self_signature =
+            ProfileChangeProof::Signature(Signature::new(SignatureType::SelfSign, self_signature));
+
+        let change_event = ProfileChangeEvent::new(event_id, changes, vec![self_signature]);
+
+        let public_kid = v.sha256(public_key.as_ref())?;
+        let public_kid = ProfileIdentifier::from_hash(public_kid);
+
+        let profile = Profile::new(
+            public_kid,
+            ProfileChangeHistory::new(vec![change_event]),
+            Vec::new(),
+            vault.clone(),
+        );
+
+        Ok(profile)
+    }
+
+    pub fn create_key(
+        &mut self,
+        label: String,
+        key_type: ProfileKeyType,
+        key_purpose: ProfileKeyPurpose,
+        attributes: Option<ProfileEventAttributes>,
+    ) -> Result<(), ockam_core::Error> {
+        let event = self.create_key_event(label, key_type, key_purpose, attributes)?;
+        self.apply_no_verification(event)
+    }
+
+    pub fn rotate_key(
+        &mut self,
+        label: String,
+        key_type: ProfileKeyType,
+        key_purpose: ProfileKeyPurpose,
+        attributes: Option<ProfileEventAttributes>,
+    ) -> Result<(), ockam_core::Error> {
+        let event = self.rotate_key_event(label, key_type, key_purpose, attributes)?;
+        self.apply_no_verification(event)
```

Comment:
Not sure why you wouldn't verify during a rotation. That seems counter-intuitive.